### ISSUE TITLE
Increasing per_page scales poorly due to WP_REST_Posts_Controller::get_item_schema calls

### ISSUE THREAD

**Issue by 27pchrisl** [WP-API/WP-API#2558]
Hi,

Whilst profiling our application, I found that as you increase per_page on a posts request there is a subsequent rise in get_item_schema calls. These calls are quite intensive (particularly if there is extra schema being added in a custom post type).

Here are some stats from kcachegrind showing that the inclusive cost of get_item_schema() calls trends upwards without caching, but with caching the inclusive cost is lower, and trends sharply downwards as per_page increases. The effect of caching is also significant increase in the speed of the overall HTTP request.

```
without schema caching:
per_page incl  self  called (overall time to run 5x http requests as reported by curl)
10       24.74 6.67  278    (3.577)
20       31.15 8.65  528    (5.701)
30       33.23 9.51  778    (7.199)
40       34.70 10.61 1028   (8.671)
100      37.85 10.80 2528   (18.925)

with schema caching:
per_page incl self called (overall time to run 5x http requests as reported by curl)
10       2.57 0.22 278    (2.899)
20       1.39 0.18 528    (3.946)
30       0.87 0.13 778    (4.887)
40       0.17 0.10 1028   (6.239)
100      0.28 0.04 2538   (12.072)
```

Here is the code that did the caching from a subclass:

```
  private $schema;
  public function get_item_schema() {
    if ($this->schema) {
      return $this->schema;
    }

    $this->schema = parent::get_item_schema();
    return $this->schema;
  }
```

**Comment by danielbachhuber** [comment]
Thanks for taking the time to dig into this. Were you able to identify the specific part to the method causing the slowness?

**Comment by 27pchrisl** [comment]
Apparently it's the number of calls to __(), which only takes 33us per call (if I'm reading this right) but it reached up to 90,000 calls from the ~2500 times get_item_schema was called.

**Comment by BE-Webdesign** [comment]
By looking for the schema property if you were doing internal requests for the same endpoint with wildly different request arguments, for instance in batch processing, you would get the improper schema returned to you because the schema would be tied to the same controller class object.  So there is some discussion that needs to happen around how this would be implemented.

**Comment by BE-Webdesign** [comment]
When we have support for batch queries we can add in a conditional to prevent the "caching" of the schema property.  On batch queries you will hit the same controller object and may need a different schema depending on the request you will be returned the wrong request. Depending on the order of calls in the batch query it could leak private information that shouldn't be leaked.  I propose using a `REST_BATCH_QUERY` constant to be defined when batch requests are happening.

**Comment by rmccue** [comment]
> ~2500 times get_item_schema was called

That sounds like a bug; not sure how it's getting this many calls. How many posts per page is this for?

**Comment by 27pchrisl** [comment]
See the left hand column of the table I posted at the top, it's between 10 and 100 per_page. I don't _think_ we are doing anything unusual on top the API to cause this behaviour (eg, weird filters causing too much load).

**Comment by BE-Webdesign** [comment]
Yeah I just profiled this too, something wrong is going on.  For a regular `GET` Request to posts `get_item_schema()` was called 46 times.  When you add more items to the response you get even more calls.  The culprits are `prepare_item_for_response()`, `get_object_type()`, and `filter_response_by_context()`. The performance impact is pretty huge actually, it takes up the bulk of a request.

**Comment by rmccue** [comment]
I dug into this issue; you're right that `get_item_schema` is scaling up linearly, and we can avoid that, but I'm not seeing it take up the majority of the call time at all. In fact, `get_item_schema` doesn't even appear, even with a sampling rate of 1ms in the flamegraph tool I'm using; the WP core code to add responsive images is much slower.

I suspected it might have been a custom language, but I installed Australian English to test, and there was no material change either with or without caching.

We can cache the schema as an object property, but it doesn't appear that we need to. I'm also only seeing `get_item_schema` called 500 times on a page with 100 posts (i.e. 5x posts), which makes much more sense to me than 2.5k times.

Is there something special with your set up? I'd love to replicate this if I can.

**Comment by BE-Webdesign** [comment]
> We can cache the schema as an object property, but it doesn't appear that we need to. I'm also only seeing get_item_schema called 500 times on a page with 100 posts (i.e. 5x posts), which makes much more sense to me than 2.5k times.
> 
> Is there something special with your set up? I'd love to replicate this if I can.

Yeah it doesn't get anywhere near 2,500 for me either; 316 calls for 100 per_page.  But the response time takes up a lot.  I think he meant it was called 2.5k times for 5x http requests so if you had 500 calls then that would be 2.5k over the course of 5 http requests, which calls `__()` about 90,000 times.

I'm using cachegrind, maybe that is not an accurate method, but over 100 runs the calls to `get_item_schema()` take up about 30% of the total processing time, when at 100 items per page.  I think we should definitely fix this, not sure if storing the schema as an object property is the best method either, but that would work.

I think potentially we would want to pass in the schema as a parameter to the three functions that are calling schema for each item.  This way we can call get_item_schema() a lot less per request.  That might not be logistically very practical though.

If this is a won't fix though, I would still understand as the API is pretty fast anyways.

**Comment by rmccue** [comment]
I wonder if this is a particular oddity in cachegrind, as I'm definitely not seeing this sort of time with the sampling method.

Let's fix anyway.

**Comment by BE-Webdesign** [comment]
I set this up with Blackfire and it shows pretty much the same stuff as the Xdebug profiler; same thing goes for XHProf. It could be the server I test against is slower on these calls for whatever reason.
